src/core/components/files.js

```diff
@@ -72,6 +74,22 @@ function normalizeContent (content) {
   })
 }
 
+function pinFile (self, opts, file, cb) {
+  // since adding paths like `directory/filename` automatically
+  // adds the directory as well as the file, we can just pin the target file
+  // and all parent dirs will be pinned indirectly
+  const pin = 'pin' in opts ? opts.pin : true
+  const isTargetFile = !file.path.includes('/')
+  const shouldPin = pin && isTargetFile && !opts.onlyHash
+  if (shouldPin) {
+    self.pin.add(file.hash, (err) => {
+      cb(err, file)
+    })
+  } else {
+    cb(null, file)
+  }
+}
```

Comment:
If I'm reading correctly, you are potentially calling `pin add --recursive` for every single file and directory that gets added in a single `ipfs files add`. This is excessive and will drag the process down.